src/ngraph/runtime/cpu/kernel/dropout.hpp

```diff
@@ -0,0 +1,85 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <random>
+
+#include "ngraph/shape.hpp"
+#include "ngraph/state/rng_state.hpp"
+
+namespace ngraph
+{
+    namespace runtime
+    {
+        namespace cpu
+        {
+            namespace kernel
+            {
+                // Note: this kernel is for doing upscale in train
+                template <typename T>
+                void generate_dropout(T* input,
+                                      T* out0,
+                                      T* out1_mask,
+                                      size_t count,
+                                      bool training,
+                                      const double value)
+                {
+                    if (training)
+                    {
+                        double dropout_prob = 1 - value;
+                        size_t nthr = 28, nelems = count;
+                        size_t chunk_size = (nelems + nthr - 1) / nthr;
+
+#pragma omp parallel num_threads(nthr)
```

Comment:
I don't think you get any randomness in the mask that gets generated.
Here is my thinking.
Random number sequence is seeded with omp thread number. So the same thread will get the same random sequence every kernel invocation. The kernel is trying to parallelize mask generation but if some openmp affinity scheduling will result in identical combined random sequence for mask generation every invocation. This sequence is then passed to uniform int distribution. The result would be the same 0, 1 mask generated every kernel invocation.

Reply:
I see. let me fix it.

Reply:
The seed should be determined from the op. Maybe op seed plus i where i is the chunk the thread is working on. The idea is that if you set the seed to a value you should get the same random values each time you run so that you can reproduce results or debug.

Reply:
In new commit, getting the seed from the op. It will be independent of tid. Initializing each thread with same seed and then calling `msr.discard()` to advance the state to necessary state.
This way the parallel threads will generate same random numbers as though one would generate in serial.

Reply:
"This way the parallel threads will generate same random numbers as though one would generate in serial." -> This is true. But you would still get the same mask every time. vmsr[tid]'s state does not get updated. 
Check this mini example.
`
#include <vector>
#include <random>
#include <iostream>
#include <functional>

size_t nthr = 8;
size_t seed = 0;
size_t chunk_size = 10;
size_t rpt = 3;

std::function<void(size_t, size_t, size_t)> lrnd;

void use(const std::vector<std::minstd_rand> &vmsr, size_t i, size_t c, size_t nt)
//void use(std::vector<std::minstd_rand> vmsr, size_t i, size_t c, size_t nt)
{
        auto rnd = vmsr[i];
        for (size_t m = 0 ; m < c; m++)
        {
            std::cout << rnd() << " ";
        }
        //rnd.discard((nt - 1) * c);
        //vmsr[i] = rnd;
}

void setup()
{
    std::vector<std::minstd_rand> vmsr(nthr);
    for (size_t i = 0; i < nthr; i++)
    {
         vmsr[i].seed(seed);
         vmsr[i].discard(i * chunk_size);
    }
    auto myrnd = [&, vmsr] (size_t i, size_t c, size_t nt)
    {
        use(vmsr, i, c, nt);
    };
    lrnd = myrnd;
}

int main()
{
    setup();
    std::minstd_rand jmsr;
    jmsr.seed(0);
    for (size_t i = 0; i < rpt; i++)
    {
        std::cout << "Iter " << i << ": " << std::endl;
        for (size_t j = 0; j < chunk_size * nthr; j++)
        {
            std::cout << jmsr() << " ";
        }
        std::cout << std::endl;
    }
    std::cout << std::endl;
    for (size_t k = 0 ; k < rpt; k++)
    {
        std::cout << "Iter " << k << ": " << std::endl;
        for (size_t n = 0 ; n < nthr; n++)
             lrnd(n, chunk_size, nthr);
        std::cout << std::endl;
    }
    std::cout << std::endl;
    return 0;
}
`

Reply:
That's right. However, the frameworks, especially PDPD, are not maintaining a state either. So, if the framework passes the same seed, then we will get same mask.
In my implementation of dropout, I am trying to be close to this PDPD native implementation so it converges and performs like native PDPD BERT : https://github.com/NervanaSystems/ngraph-paddle/blob/14d88829b386c9f7601788c5539c08326dcbe2fe/paddle/fluid/operators/dropout_op.h#L58-L78
However, we also looked at other frameworks and they don't maintain state MxNet : https://github.com/apache/incubator-mxnet/blob/503c7509db67c43e0afe8aec0e006b0cacaf7128/src/operator/nn/dropout-inl.h#L90-L112
Tensorflow : https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/ops/nn_ops.py#L3044-L3048